File: waveform-analyzer/python/waveform_report/stats.py
from __future__ import annotations

from typing import Any

import numpy as np

NAN_RATIO_LIMIT = 0.001


def valid(x: np.ndarray) -> np.ndarray:
    return x[np.isfinite(x)]


def channel_stats(values: list[float | None] | np.ndarray) -> dict[str, float | None]:
    arr = valid(np.asarray(values, dtype=float))
    empty = {
        "peak": None,
        "min": None,
        "peakPeak": None,
        "average": None,
        "rms": None,
        "crestFactor": None,
        "ripple": None,
        "thd": None,
    }
    if arr.size < 1:
        return empty
    peak = float(np.max(arr))
    mn = float(np.min(arr))
    avg = float(np.mean(arr))
    rms = float(np.sqrt(np.mean(arr * arr)))
    crest = None
    if rms > 0:
        crest = float(max(abs(peak), abs(mn)) / rms)
    ripple = None
    if abs(avg) > 0:
        ripple = float((peak - mn) / abs(avg) * 100.0)
    return {
        "peak": peak,
        "min": mn,
        "peakPeak": float(peak - mn),
        "average": avg,
        "rms": rms,
        "crestFactor": crest,
        "ripple": ripple,
        "thd": None,
    }


def imbalance_percent(rms_values: list[float | None]) -> float | None:
    vals = [v for v in rms_values if v is not None and np.isfinite(v)]
    if len(vals) != 3:
        return None
    mean = float(sum(vals) / 3.0)
    if mean == 0:
        return None
    return float(max(abs(v - mean) for v in vals) / mean * 100.0)


def nan_ratio(values: list[float | None] | np.ndarray, expected: int) -> float:
    arr = np.asarray(values, dtype=float)
    n = max(expected, arr.size)
    if n == 0:
        return 1.0
    invalid = n - int(np.isfinite(arr).sum())
    invalid += max(0, expected - arr.size)
    return invalid / n


def default_channels() -> list[dict[str, str | None]]:
    return [
        {"id": "Uu", "group": "voltage", "pairId": "phase-U", "unit": "V"},
        {"id": "Vv", "group": "voltage", "pairId": "phase-V", "unit": "V"},
        {"id": "Ww", "group": "voltage", "pairId": "phase-W", "unit": "V"},
        {"id": "Iu", "group": "current", "pairId": "phase-U", "unit": "A"},
        {"id": "Iv", "group": "current", "pairId": "phase-V", "unit": "A"},
        {"id": "Iw", "group": "current", "pairId": "phase-W", "unit": "A"},
        {"id": "speed", "group": "motor", "pairId": None, "unit": "rpm"},
        {"id": "load", "group": "motor", "pairId": None, "unit": "%"},
    ]


def iter_series(payload: dict[str, Any]) -> list[tuple[str, str, list]]:
    out: list[tuple[str, str, list]] = []
    for group in payload:
        if group in {"sampleCount", "samplingRate", "units", "channels", "warnings", "stats"}:
            continue
        block = payload.get(group)
        if not isinstance(block, dict):
            continue
        for ch_id, series in block.items():
            if isinstance(series, list):
                out.append((group, ch_id, series))
    return out


def build_waveform_document(
    *,
    sample_count: int,
    sampling_rate: float,
    groups: dict[str, dict[str, list]],
    units: dict[str, str] | None = None,
    channels: list[dict[str, str | None]] | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "sampleCount": int(sample_count),
        "samplingRate": float(sampling_rate),
        "units": units
        or {"voltage": "V", "current": "A", "speed": "rpm", "load": "%"},
        "channels": channels or default_channels(),
        "warnings": [],
        "stats": {"full": {}},
    }
    payload.update(groups)

    full: dict[str, Any] = {}
    voltage_rms: list[float | None] = []
    current_rms: list[float | None] = []

    for group, ch_id, series in iter_series(payload):
        actual = len(series)
        if actual != sample_count:
            payload["warnings"].append(
                {
                    "channel": ch_id,
                    "expected": sample_count,
                    "actual": actual,
                    "message": f"{ch_id} length {actual} < {sample_count}"
                    if actual < sample_count
                    else f"{ch_id} length {actual} > {sample_count}, truncated in view",
                }
            )
        sliced = series[:sample_count]
        st = channel_stats(sliced)
        full[ch_id] = st
        too_many_missing = nan_ratio(sliced, sample_count) > NAN_RATIO_LIMIT
        if group == "voltage" and ch_id in {"Uu", "Vv", "Ww"}:
            voltage_rms.append(None if too_many_missing else st["rms"])
        if group == "current" and ch_id in {"Iu", "Iv", "Iw"}:
            current_rms.append(None if too_many_missing else st["rms"])

    full["voltageImbalance"] = imbalance_percent(voltage_rms)
    full["currentImbalance"] = imbalance_percent(current_rms)
    payload["stats"]["full"] = full
    return payload

File: waveform-analyzer/viewer/src/waveform/stats.ts
import { NAN_RATIO_LIMIT, type ChannelStats } from "./types"

export function channelStats(values: Array<number | null | undefined>): ChannelStats {
  const arr = values.filter((v): v is number => typeof v === "number" && Number.isFinite(v))
  const empty: ChannelStats = {
    peak: null,
    min: null,
    peakPeak: null,
    average: null,
    rms: null,
    crestFactor: null,
    ripple: null,
    thd: null,
  }
  if (arr.length < 1) return empty
  let peak = -Infinity
  let min = Infinity
  let sum = 0
  let sumSq = 0
  for (const v of arr) {
    if (v > peak) peak = v
    if (v < min) min = v
    sum += v
    sumSq += v * v
  }
  const average = sum / arr.length
  const rms = Math.sqrt(sumSq / arr.length)
  return {
    peak,
    min,
    peakPeak: peak - min,
    average,
    rms,
    crestFactor: rms > 0 ? Math.max(Math.abs(peak), Math.abs(min)) / rms : null,
    ripple: Math.abs(average) > 0 ? ((peak - min) / Math.abs(average)) * 100 : null,
    thd: null,
  }
}

export function imbalancePercent(rmsValues: Array<number | null>): number | null {
  const vals = rmsValues.filter((v): v is number => v != null && Number.isFinite(v))
  if (vals.length !== 3) return null
  const mean = vals.reduce((a, b) => a + b, 0) / 3
  if (mean === 0) return null
  return (Math.max(...vals.map((v) => Math.abs(v - mean))) / mean) * 100
}

export function nanRatio(values: Float64Array, expected: number): number {
  const n = Math.max(expected, values.length)
  if (n === 0) return 1
  let finite = 0
  for (let i = 0; i < values.length; i++) {
    if (Number.isFinite(values[i])) finite++
  }
  return (n - finite) / n
}

export { NAN_RATIO_LIMIT }
